### PR TITLE
Fixes underscores when displaying commands and plugin names

### PR DESCRIPTION
**Pull request by Kleptobismol**

### CODE REVIEW

File: q2cli/cli.py [hunk 1]
```diff
@@ -25,17 +25,24 @@ class QiimeCLI(click.MultiCommand):
     _builtin_commands = {'tools': q2cli._tools.tools,
                          'info': q2cli._info.info}
 
+    @property
+    def _name_map(self):
+        return {k.replace('_', '-'): k for k in self._plugin_manager.plugins}
```
Comment (gregcaporaso) [inline q2cli/cli.py]: Can you replace ``k.replace('_', '-')`` with ``_name_to_command(k)``, so you're not duplicating the renaming logic here?

File: q2cli/cli.py [hunk 2]
```diff
@@ -25,17 +25,24 @@ class QiimeCLI(click.MultiCommand):
...
+
     def list_commands(self, ctx):
-        plugins = list(sorted(self._plugin_manager.plugins.keys()))
+        plugins = list(sorted(self._name_map.values()))
+        accepted_plugins = []
         for idx, key in enumerate(plugins):
             plugin = self._plugin_manager.plugins[key]
-            if not plugin.methods and not plugin.visualizers:
-                del plugins[idx]
+            if plugin.methods and plugin.visualizers:
```
Comment (gregcaporaso) [inline q2cli/cli.py]: This should be ``or`` (since we want there to be at least one)?

File: q2cli/cli.py [hunk 3]
```diff
@@ -25,32 +25,45 @@ class QiimeCLI(click.MultiCommand):
...
-        elif name in self._plugin_manager.plugins:
+        name = self._name_map[name]
+        if name in self._plugin_manager.plugins:
             plugin = self._plugin_manager.plugins[name]
 
             class PluginCommand(click.MultiCommand):
+                @property
+                def _name_map(self):
+                    return {_name_to_command(k) : k for k in
+                            list(plugin.methods.keys()) +
```
Comment (gregcaporaso) [inline q2cli/cli.py]: Drop you can remove ``keys()`` here.

File: q2cli/cli.py [hunk 4]
```diff
@@ -25,32 +25,45 @@ class QiimeCLI(click.MultiCommand):
     _builtin_commands = {'tools': q2cli._tools.tools,
                          'info': q2cli._info.info}
 
+    @property
+    def _name_map(self):
+        plugins = list(sorted(self._plugin_manager.plugins))
```
Comment (jairideout) [inline q2cli/cli.py]: Remove this line

File: q2cli/cli.py [hunk 5]
```diff
@@ -25,32 +25,45 @@ class QiimeCLI(click.MultiCommand):
...
+        accepted_plugins = []
+        for plugin_name in plugins:
```
Comment (jairideout) [inline q2cli/cli.py]: Change to `for plugin_name, plugin in self._plugin_manager.plugins.items():`

File: q2cli/cli.py [hunk 6]
```diff
@@ -25,32 +25,45 @@ class QiimeCLI(click.MultiCommand):
     _builtin_commands = {'tools': q2cli._tools.tools,
                          'info': q2cli._info.info}
 
+    @property
+    def _name_map(self):
+        plugins = list(sorted(self._plugin_manager.plugins))
+        accepted_plugins = []
```
Comment (jairideout) [inline q2cli/cli.py]: Change this to `name_map = {}` and populate it with converted name to original name, then return `name_map` at end of function.

File: q2cli/cli.py [hunk 7]
```diff
@@ -25,32 +25,45 @@ class QiimeCLI(click.MultiCommand):
...
+        return {_name_to_command(k): k for k in accepted_plugins}
+
     def list_commands(self, ctx):
-        plugins = list(sorted(self._plugin_manager.plugins.keys()))
-        for idx, key in enumerate(plugins):
-            plugin = self._plugin_manager.plugins[key]
-            if not plugin.methods and not plugin.visualizers:
-                del plugins[idx]
         builtins = list(sorted(self._builtin_commands.keys()))
+        plugins = list(self._name_map.keys())
```
Comment (jairideout) [inline q2cli/cli.py]: You don't need to call `.keys()`when iterating over a `dict`.

Comment (jairideout) [inline q2cli/cli.py]: Sort these plugin commands

File: q2cli/cli.py (hunk 5 above)
Comment (Kleptobismol) [inline q2cli/cli.py]: Fixed

File: q2cli/cli.py (hunk 4 above)
Comment (Kleptobismol) [inline q2cli/cli.py]: Fixed

File: q2cli/cli.py (hunk 6 above)
Comment (Kleptobismol) [inline q2cli/cli.py]: Fixed

File: q2cli/cli.py [hunk 8]
```diff
@@ -25,32 +25,45 @@ class QiimeCLI(click.MultiCommand):
...
         commands = builtins + plugins
         return commands
 
     def get_command(self, ctx, name):
         if name in self._builtin_commands:
             return self._builtin_commands[name]
-        elif name in self._plugin_manager.plugins:
+        name = self._name_map[name] if name in self._name_map else None
```
Comment (jairideout) [inline q2cli/cli.py]: Change this check to:

```python
name_map = self._name_map
if name in name_map:
    plugin_name = name_map[name]
    plugin = self._plugin_manager.plugins[plugin_name]
    ...
else:
    return None
```

File: q2cli/cli.py (hunk 7 above)
Comment (Kleptobismol) [inline q2cli/cli.py]: Fixed

File: q2cli/cli.py [hunk 9]
```diff
@@ -25,32 +25,45 @@ class QiimeCLI(click.MultiCommand):
...
+                @property
+                def _name_map(self):
+                    return {_name_to_command(k) : k for k in
+                            list(plugin.methods) +
+                            list(plugin.visualizers)}
 
                 def list_commands(self, ctx):
                     # the cli currently doesn't differentiate between methods
                     # and visualizers.
                     commands = list(plugin.methods.keys()) + \
```
Comment (jairideout) [inline q2cli/cli.py]: You can replace this entire method with `return sorted(self._name_map)`

File: q2cli/cli.py (hunk 8 above)
Comment (Kleptobismol) [inline q2cli/cli.py]: Fixed

File: q2cli/cli.py [hunk 10]
```diff
@@ -25,32 +25,45 @@ class QiimeCLI(click.MultiCommand):
...
                                list(plugin.visualizers.keys())
-                    return sorted(commands)
+                    return sorted([_name_to_command(n) for n in commands])
 
                 def get_command(self, ctx, name):
+                    name = self._name_map[name] if name in self._name_map else name
```
Comment (jairideout) [inline q2cli/cli.py]: Change this to:

```python
name_map = self._name_map
if name not in name_map:
    click.echo(("%s is not a valid...")
    ...
action_name = name_map[name]
if action_name in plugin.methods:
    return _build_method_command(
        name, plugin.methods[action_name])
else:
    return _build_visualizer_command(
        name, plugin.visualizers[action_name])
```

Comment (Kleptobismol) [inline q2cli/cli.py]: Fixed

**Comment by jairideout** [comment]
Thanks @Kleptobismol!!!

**Pull request merged**